### PR TITLE
chore: add README to clock testing

### PR DESCRIPTION
**Pull request by spikecurtis**
Adds README with some draft content explaining why the library exists.  Will be most relevant when we spin out into a standalone library.

### CODE REVIEW

**Comment by spikecurtis** [comment]
* **#13583** <a href="https://app.graphite.dev/github/pr/coder/coder/13583?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: clock/README.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+Writing good unit tests for components and functions that use the `time` package is difficult, and
+several open source libraries exist.  In building Quartz, we took some inspiration from
+
+ * [github.com/benbjohnson/clock](https://github.com/benbjohnson/clock)
+ * Tailscale's [tstest.Clock](https://github.com/coder/tailscale/blob/main/tstest/clock.go)
+ * [github.com/aspenmesh/tock](https://github.com/aspenmesh/tock)
+
+Quartz shares the high level design of a `Clock` interface that closely resembles the functions in
+the `time` standard library, and a "real" clock passes thru to the standard libary in production,
```
Comment (johnstcn) [inline clock/README.md:17]:
```suggestion
the `time` standard library, and a "real" clock passes through to the standard library in production,
```

File: clock/README.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
+# Quartz
+A Go time testing library for writing deterministic unit tests
+
+_Note: Quartz is the name I'm targeting for the standalone open source project when we spin this
+out._
+
+## Why another time testing library?
+
+Writing good unit tests for components and functions that use the `time` package is difficult, and
+several open source libraries exist.  In building Quartz, we took some inspiration from
```
Comment (johnstcn) [inline clock/README.md:10]:
```suggestion
Writing good unit tests for components and functions that use the `time` package is difficult, even though
several open source libraries exist.  In building Quartz, we took some inspiration from
```

File: clock/README.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+This affords the mock `Clock` the ability to explicitly know when processing of a tick is finished
+because it's wrapped in the function passed to `TickerFunc` (`do()` in this example).
+
+In Quartz, when you advance the clock, you are returned an object you can `Wait()` on to ensure all
+ticks and timers triggered are finished. This solves the first race condition in the example.
+
+(As an aside, we still support a traditional standard library-style `Ticker`, if you want to keep
+your code as close as possible to the standard library, or if you need to use the channel in a
+larger `select` block, but in that case you'll have to find some other mechanism to sync tick
+processing to your test code.)
```
Comment (johnstcn) [inline clock/README.md:106]: Run-on sentence, consider breaking up.

File: clock/README.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+ * [github.com/aspenmesh/tock](https://github.com/aspenmesh/tock)
+
+Quartz shares the high level design of a `Clock` interface that closely resembles the functions in
+the `time` standard library, and a "real" clock passes thru to the standard libary in production,
+while a mock clock gives precise control in testing.
+
+Our high level goal is to write unit tests that
+
+1. execute quickly
+2. don't flake
```
Comment (johnstcn) [inline clock/README.md:23]: suggestion: avoid colloquialism
```suggestion
2. execute reliably
```

File: clock/README.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+3. are straightforward to write and understand
+
+For several reasons, this is a tall order when it comes to code that depends on time, and we found
+the existing libraries insufficient for our goals.
+
+For tests to execute quickly without flakes, we want to focus on _determinism_: the test should run
```
Comment (johnstcn) [inline clock/README.md:29]:
```suggestion
For tests to execute quickly and reliably, we want to focus on _determinism_: the test should run
```

**Review by johnstcn — APPROVED** [review]
Some minor language nits, otherwise nice writeup!

File: clock/README.md (hunk 1 above)
Comment (spikecurtis) [reply clock/README.md:17]: https://www.youtube.com/watch?v=FXBHY7uco0Y

File: clock/README.md (hunk 4 above)
Comment (spikecurtis) [reply clock/README.md:23]: While arguably "jargon", I think flaky is more precise. It refers to tests that succeed or fail due to uncontrolled randomness.

A test might not be reliable for other reasons, such as not actually testing or asserting the thing is purports to.

**Pull request merged**